### PR TITLE
add Format-like printers

### PR DESCRIPTION
**Pull request by vbmithr**
Needed some support for `Format`. What do you think about it?

### CODE REVIEW

File: src/ANSITerminal.mli [hunk 1]
```diff
@@ -90,23 +90,38 @@ val print_string : style list -> string -> unit
...
-(** Same as {!printf} but prints the result on [stderr]. *)
-
-val sprintf : style list -> ('a, unit, string) format -> 'a
-(** Same as {!printf} but returns the result in a string.  This only
-    works on ANSI compliant terminals — for which escape sequences are
-    used — and not under Windows — where system calls are required.
-    On Windows, it is identical to the standard [sprintf]. *)
-
+(** Like [print_string] but prints on the standard error. *)
+
```
Comment (Chris00) [inline src/ANSITerminal.mli:94]: The above two lines are not part of your changes  — keep them as they were.

File: src/ANSITerminal.mli [hunk 2]
```diff
@@ -90,23 +90,38 @@ val print_string : style list -> string -> unit
...
+  val sprintf : style list -> ('a, unit, string) format -> 'a
+  (** Same as {!printf} but returns the result in a string.  This only
+      works on ANSI compliant terminals — for which escape sequences are
+      used — and not under Windows — where system calls are required.
+      On Windows, it is identical to the standard [sprintf]. *)
+
+  val fprintf :
+    style list -> out_channel -> ('a, unit, string, unit) format4 -> 'a
+end
```
Comment (Chris00) [inline src/ANSITerminal.mli:113]: Please document this function.

File: src/ANSITerminal.mli [hunk 3]
```diff
@@ -90,23 +90,38 @@ val print_string : style list -> string -> unit
...
+
+module Format : sig
```
Comment (Chris00) [inline src/ANSITerminal.mli:115]: All the functions of this module and the module itself need to be documented.

File: src/ANSITerminal_unix.ml [hunk 4]
```diff
@@ -175,26 +175,28 @@ let style_to_string = function
   | Background White -> "47"
   | Background Default -> "49"
 
-
-let print_with pr ~tty style txt =
+let print_with style txt oc =
```
Comment (Chris00) [inline src/ANSITerminal_unix.ml:178]: I do not mind you modify earlier code as part of your PR but could you please split it in two commits: one for existing code modifications — including moving printf functions to their own sub-module — and one adding Format support.

File: src/ANSITerminal_unix.ml [hunk 5]
```diff
@@ -175,26 +175,28 @@ let style_to_string = function
...
+    output_string oc "\027[";
+    output_string oc (String.concat ";" (List.map style_to_string style));
+    output_string oc "m";
   );
-  pr txt;
-  if tty && !autoreset then pr "\027[0m"
+  output_string oc txt;
+  if tty && !autoreset then output_string oc "\027[0m"
+
+let ppf_print_with style txt ppf =
```
Comment (Chris00) [inline src/ANSITerminal_unix.ml:188]: This function should be in the `Format` sub-module IMHO.

File: src/ANSITerminal_win.ml [hunk 6]
```diff
@@ -113,11 +113,20 @@ let print ch styles txt =
...
-let printf style = kprintf (print_string style)
-
-let eprintf style = ksprintf (prerr_string style)
-
-let sprintf style = sprintf
+module Printf = struct
+  let printf style = kprintf (print_string style)
+  let eprintf style = ksprintf (prerr_string style)
+  let sprintf _style = sprintf
+  let fprintf _ _ _ = invalid_arg "not implemented"
```
Comment (Chris00) [inline src/ANSITerminal_win.ml:120]: Need to be implemented.

File: src/ANSITerminal_win.ml [hunk 7]
```diff
@@ -113,11 +113,20 @@ let print ch styles txt =
...
+end
+
+module Format = struct
```
Comment (Chris00) [inline src/ANSITerminal_win.ml:123]: Functions of this module need to be implemented on Windows as well — make it a separated commit if you wish.

**Review by Chris00 — CHANGES_REQUESTED** [review]

**Comment by vbmithr** [comment]
Thanks for the comments. Closing for now, might re-open and work on it later.

**Pull request closed without merging**